### Mac Cassin MSDS 460 Fall 2023
## The Diet Problem Revisited (version 6)

Developed by Thomas W. Miller. Revised September 2, 2023.

Introduced by Stigler (1945) and discussed by Dantzig (1990), the diet problem is a classic problem in constrained optimization often used to introduce linear programming concepts. The diet problem has been studied extensively through the years (van Dooren 2018).

For this assignment, you need to construct a personalized diet using current recommended dietary allowances from the U.S. Food and Drug Administration, updated to account for recent research on sodium intake and health (Mente, O'Donnell, and Yusuf 2021). Interested in learning more about nutrition and healthy living? Check out Nutrients, an open-access journal, at its home page https://www.mdpi.com/journal/nutrientsLinks to an external site. .

The constraints for this linear programming problem, should consider seven components of nutrition, as shown in the following table:

Component

Max/Min

Amount and measure

| Sodium | Maximum | 5,000 milligrams (mg) |

| Energy | Minimum | 2,000 Calories (kilocalories, kcal) |

| Protein | Minimum | 50 grams (g) |

| Vitamin D | Minimum | 20 micrograms (mcg) |

| Calcium | Minimum | 1,300 milligrams (mg) |

| Iron | Minimum | 18 milligrams (mg) |

| Potassium | Minimum | 4,700 milligrams (mg) |

Set this up as a standard linear programming problem with decision variables taking any non-negative values. In other words, partial servings are permitted.  

Nutrition labels on packaged foods should contain information about these eight components of nutrition along with other components. Each of the components represents a constraint in the linear programming problem you are developing.

To adapt the problem to your personal diet, collect nutrition facts from five packaged food items in your household. Use packaged foods that are part of your normal diet and for which you have prices. Also, ensure that the across the set of food items there are positive values for each of the eight components of nutrition. Adjust the price for each food item so that it represents one serving size, as defined on the nutrition facts label. Each food item represents a decision variable in the linear programming problem.

The goal or objective of this problem is to find the minimum-cost diet (servings of food items) that satisfies the eight nutritional requirements. Use Python Pulp to solve this linear programming problem. 

Deliverables (150 points total, 30 points for each part)

Paper (pdf file). The paper/write-up should be submitted as a pdf file (2 pages max). Think of the paper as comprising the methods and results sections of a written research report. If you like, provide a paragraph on methods and a paragraph about results for each of the five parts of this assignment.

Program code (text link to GitHub repository). Key information from the paper should also be included in the README.md markdown file of a public GitHub repository established by the student. The GitHub repository should include text files for the program code (.py extension for Python), and program output (.txt extension). Include the web address (URL) of the GitHub repository text in the posted submission for the assignment. Image files (.jpg or .png extension) should be included for the scanned food labels.

Uploads are restricted to files with pdf, md, and txt extensions.

Part 1. Provide documentation for the five packaged food items you have selected for the assignment. Photographs of the Nutrition Facts labels are sufficient. Show your price calculations for serving sizes.
Part 2. Specify the linear programming problem in standard form, showing decision variables, objective function with cost coefficients, and nutritional constraints. Describe the problem in plain English.
Part 3. Implement the linear programming problem using Python PuLP. Provide the program code and output/listing as plain text files (.py, .txt). (If you use Jupyter notebook, export the code as a plain Python text file and save the code and listing as an HTML file.)
Part 4. Describe the solution for the linear programming problem, showing units (serving sizes) for each of the five food items. What is the minimum cost solution? How much would you need to spend on food each day?
Part 5. What happens when you add additional nutritional constraints, drawing from recommendations of the U.S. Food and Drug Administration: https://www.fda.gov/media/99069/downloadLinks to an external site.. Try adding two to six nutritional constraints, such as minimum requirements for vitamin C, niacin (vitamin B3), vitamin E, vitamin B6, and magnesium, for example. Solve the revised linear programming problem. How does this change the solution? Did you have to add food items to satisfy the expanded set of constraints? How much more would you have to spend on food each day?
 

References

Dantzig, George B. 1990. “The Diet Problem.” Informs. 20:4, 43–47. Available on Course Reserves.

Food and Drug Administration, Department of Health and Human Services. 2016. Food Labeling: Revision of the Nutrition and Supplemental Facts Labels. Available at https://s3.amazonaws.com/public-inspection.federalregister.gov/2016-11867.pdfLinks to an external site.

Mente, Andrew, Martin O'Donnell, and Salim Yusuf. 2021, September. "Sodium Intake and Health: What Should We Recommend Based on the Current Evidence?" Nutrients, 13(9): 3232. Available online at https://www.mdpi.com/2072-6643/13/9/3232Links to an external site. . This is part of the September 2021 special issue of Nutrients: "Towards Better Dietary Guidelines: New Approaches Based on Recent Science."  

Stigler, George. 1945. “The Cost of Subsistence.” Journal of Farm Economics, 25:2, 303–314. Available online at https://math.berkeley.edu/~mgu/MA170/Diet.pdfLinks to an external site.

van Dooren, Corné. 2018, June. “A Review of the Use of Linear Programming to Optimize Diets, Nutritiously, Economically and Environmentally.” Frontiers in Nutrition, 4, Article 48. Available online at https://www.frontiersin.org/articles/10.3389/fnut.2018.00048/fullLinks to an external site.


I wanted to make this project as realistic as possible for myself, which means pulling all of the nutrition and cost information for multiple ingredients for two of the options I am using. I created a table with the price, number of servings per item, and number of servings per ingredient to get the cost per meal. For my two meals, "Slamon Salad" and "Curry," I then combined the cost per meal per item to get the total price for the whole meal. 
Initially, I used the nutrition label on each physical package and looked up the nutrition data for produce items. However, regarding the 5th part of this problem, only some of the nutrition labels had any additional data fields I wanted to investigate other than Cholesterol. I tried my best to source more in-depth nutrition data from online sources that were as close to, if different from, the exact match for the item I purchased. 

### Stand Alone Items
| Food          | Cost  | Servings per Purchase | Servings Per Meal | Price Per Serving |
|---------------|-------|------------------------|--------------------|-------------------|
| Eggs          | 2.89  | 18                     | 1.00               | 0.16              |
| Bread         | 7.49  | 44                     | 1.00               | 0.17              |
| Avocado       | 3.38  | 2                      | 0.25               | 0.42              |

### Salmon Salad
| Food          | Cost  | Servings per Purchase | Servings Per Meal | Price Per Serving |
|---------------|-------|------------------------|--------------------|-------------------|
| Salmon        | 13.99 | 4                      | 1                  | 3.50              |
| Beets         | 2.99  | 10                     | 2                  | 0.60              |
| Spring Mix    | 5.99  | 5                      | 1                  | 1.20              |
| Mustard       | 1.99  | 68                     | 2                  | 0.06              |
| Olive Oil     | 19.99 | 133                    | 1                  | 0.15              |
| Feta          | 7.99  | 12                     | 0.5                | 0.33              |
| Croutons      | 1.99  | 20                     | 1                  | 0.10              |
| **SUM**       |       |                        |                    | 5.93              |


### Curry 
| Food          | Cost  | Servings per Purchase | Servings Per Meal | Price Per Serving |
|---------------|-------|------------------------|--------------------|-------------------|
| Broccoli      | 1.68  | 1                      | 0.13               | 0.21              |
| Coconut Milk  | 1.99  | 5                      | 0.63               | 0.25              |
| Curry Paste G | 3.29  | 26                     | 3.25               | 0.41              |
| Bell Peppers  | 1.78  | 1                      | 0.13               | 0.22              |
| Onion         | 1.29  | 1                      | 0.13               | 0.16              |
| Coconut Oil   | 8.99  | 60                     | 1.00               | 0.15              |
| Chicken       | 12.46 | 1                      | 0.06               | 0.78              |
| Fish Sauce    | 4.59  | 13                     | 0.06               | 0.02              |
| **SUM**       |       |                        |                    | 2.20              |


I then followed the same principle for determining the nutritional content of each item. However, the serving sizes for some ingredients are easier to deal with than others. For example, I use one tablespoon of coconut oil for the chicken in the Curry. However, the nutrition data I got for vegetables and meat were weight-based. To accommodate this, I calculated the nutritional content for the entirety of that item purchased and modified the serving size to account for this. This is why you will see the raw nutrition data below range from single-serving data to much larger amounts (chicken being the best example). This was an intermediary step. I wanted to cleanly convert the nutrition data to the amount contained in a single meal. This is all sorted by further breaking down the number of servings.

That results in the following breakdowns of nutrition per ingredient:

#### Salmon Salad

| Food        | Sodium | Calories | Protein | Vitamin D | Calcium | Iron | Potassium | Cholesterol | Vitamin C | Vitamin K | Vitamin A | Folate | Nutrition Data                                                                                                       |
|-------------|--------|----------|---------|-----------|---------|------|-----------|-------------|-----------|-----------|-----------|--------|---------------------------------------------------------------------------------------------------------------------|
| Salmon      | 420    | 140      | 12      | 6         | 26      | 0    | 208       | 30          | 0         | 1.2       | 162       | 5.1    | [Link](https://www.urmc.rochester.edu/encyclopedia/content.aspx?contenttypeid=76&contentid=15087-2)             |
| Beets       | 50     | 15       | 0       | 0         | 2.5     | 0.09 | 33.5      | 0           | 0.5       | 0         | 2.5       | 6.1    | [Link](https://www.urmc.rochester.edu/encyclopedia/content.aspx?contenttypeid=76&contentid=11609-1)             |
| Spring Mix  | 60     | 20       | 2       | 0.2       | 70      | 1.5  | 290       | 0           | 19        | 124       | 120       | 0      | [Link](https://www.verywellfit.com/spring-mix-nutrition-facts-calories-and-health-benefits-4119165)            |
| Mustard     | 115    | 0        | 0       | 0         | 0       | 0.1  | 0         | 0           | 0         | 0         | 0         | 0      | [Link](https://www.recipal.com/ingredients/18704-nutrition-facts-calories-protein-carbs-fat-dijon-mustard)  |
| Olive Oil   | 0.27   | 119      | 0       | 0         | 0       | 0    | 0         | 0           | 0         | 8.1       | 0         | 0      | [Link](https://fdc.nal.usda.gov/fdc-app.html#/food-details/171413/nutrients)                                  |
| Feta        | 330    | 70       | 5       | 0         | 70      | 0    | 0         | 20          | 0         | 0.31      | 0         | 5.4    | [Link](https://www.urmc.rochester.edu/encyclopedia/content.aspx?contenttypeid=76&contentid=01019-3)           |
| Croutons    | 90     | 30       | 1       | 0         | 0       | 0    | 0         | 0           | 0         | 0.5       | 2.1       | 10     | [Link](https://www.urmc.rochester.edu/encyclopedia/content.aspx?contenttypeid=76&contentid=18243-1)          |


#### Curry 
| Food          | Sodium | Calories | Protein | Vitamin D | Calcium | Iron | Potassium | Cholesterol | Vitamin C | Vitamin K | Vitamin A | Folate | Nutrition Data                                                                                                       |
|---------------|--------|----------|---------|-----------|---------|------|-----------|-------------|-----------|-----------|-----------|--------|---------------------------------------------------------------------------------------------------------------------|
| Broccoli      | 180    | 180      | 14.4    | 0         | 252     | 3.8  | 1668.5    | 0           | 454.8     | 518.1     | 3176.6    | 321.2  | [Link](https://www.urmc.rochester.edu/encyclopedia/content.aspx?contenttypeid=76&contentid=11090-1)             |
| Coconut Milk  | 20     | 150      | 1       | 0         | 0       | 0.6  | 120       | 0           | 0.76      | 0.08      | 0.00      | 10.66  | [Link](https://www.urmc.rochester.edu/encyclopedia/content.aspx?contenttypeid=76&contentid=12118-1)            |
| Curry Paste G | 100    | 5        | 0       | 0         | 0       | 0    | 0         | 0           | 0         | 0         | 0         | 0      | [Link](https://www.recipal.com/ingredients/9104-nutrition-facts-calories-protein-carbs-fat-thai-kitchen-green-curry-paste)  |
| Bell Peppers  | 4.6    | 64       | 2       | 0         | 20      | 1    | 378.5     | 0           | 284       | 12.02     | 7681      | 44.16  | [Link](https://www.urmc.rochester.edu/encyclopedia/content.aspx?contenttypeid=76&contentid=11821-2)          |
| Onion         | 7.5    | 75       | 2.1     | 0         | 43      | 0.4  | 274       | 0           | 13.9      | 0.8       | 0         | 35.72  | [Link](https://www.nutritionvalue.org/Onions%2C_raw_nutritional_value.html?size=188+g)                      |
| Coconut Oil   | 0      | 120      | 0       | 0         | 0       | 0.01 | 0         | 0           | 0         | 0.07      | 0         | 0      | [Link](https://www.urmc.rochester.edu/encyclopedia/content.aspx?contenttypeid=76&contentid=04047-1)           |
| Chicken       | 425    | 1135     | 215     | 0         | 47.5    | 3.5  | 3156.3    | 690         | 0         | 79.8      | 285       | 85.5   | [Link](https://fdc.nal.usda.gov/fdc-app.html#/food-details/171077/nutrients)                                  |
| Fish Sauce    | 1480   | 10       | 2       | 0         | 7.7     | 0.1  | 51.8      | 0           | 0         | 0         | 0         | 0      | [Link](https://www.webstaurantstore.com/documents/nutrition/sauce_fish_premium_62366_oz_thai_kitchen.pdf)   |


#### Stand Alone Items
| Food    | Sodium | Calories | Protein | Vitamin D | Calcium | Iron | Potassium | Cholesterol | Vitamin C | Vitamin K | Vitamin A | Folate | Nutrition Data                                                                                                       |
|---------|--------|----------|---------|-----------|---------|------|-----------|-------------|-----------|-----------|-----------|--------|---------------------------------------------------------------------------------------------------------------------|
| Eggs    | 70     | 70       | 6       | 1         | 30      | 0.9  | 70        | 185         | 0         | 0.143     | 239.1     | 22.3   | [Link](https://www.urmc.rochester.edu/encyclopedia/content.aspx?contenttypeid=76&contentid=01123-1)             |
| Bread   | 170    | 110      | 5       | 0         | 26      | 1    | 100       | 0           | 0         | 0         | 0         | 0      | [Link](https://www.eatthismuch.com/food/nutrition/organic-thin-sliced-21-whole-grains-and-seeds-brea,496247/)  |
| Avocado | 14     | 322      | 4       | 0         | 24      | 1.1  | 974.9     | 0           | 20.1      | 42        | 293       | 162    | [Link](https://www.urmc.rochester.edu/encyclopedia/content.aspx?contenttypeid=76&contentid=09037-4)           |



The final step was converting the total number of servings needed per meal and dividing that by the number of Meals created. The amount of ingredients for the Curry usually makes around six servings. This finally gets us to the total cost per meal with the nutritional content per meal, which we will use for this optimization problem. 

### Breakdowns for optimization 

| Food/Meal    | Cost/Serving | Sodium | Calories | Protein | Vitamin D | Calcium | Iron | Potassium | Cholesterol | Vitamin C | Vitamin K | Vitamin A | Folate |
|--------------|--------------|--------|----------|---------|-----------|---------|------|-----------|-------------|-----------|-----------|-----------|--------|
| Eggs         | 0.16         | 70.0   | 70.0     | 6.0     | 1.0       | 30.0    | 0.9  | 70.0      | 185.0       | 0.0       | 0.1       | 239.1     | 22.3   |
| Bread        | 0.17         | 170.0  | 110.0    | 5.0     | 0.0       | 26.0    | 1.0  | 100.0     | 0           | 0         | 0         | 0         | 0      |
| Avocado      | 0.42         | 3.5    | 80.5     | 1.0     | 0.0       | 6.0     | 0.3  | 243.7     | 0.0         | 5.0       | 10.5      | 73.3      | 40.5   |
| Curry        | 2.20         | 480.3  | 342.3    | 16.6    | 0.0       | 42.7    | 1.3  | 570.4     | 43.1        | 94.6      | 71.5      | 1375.0    | 62.1   |
| Salmon Salad | 5.93         | 1185.0 | 255.0    | 17.5    | 6.2       | 131.0   | 1.5  | 498.0     | 40          | 20        | 125.855   | 289.1     | 30     |



**Objective Function:**

Minimize <code>Z = c<sub>1</sub>x<sub>1</sub> + c<sub>2</sub>x<sub>2</sub> + c<sub>3</sub>x<sub>3</sub> + c<sub>4</sub>x<sub>4</sub> + c<sub>5</sub>x<sub>5</sub> </code>

**Where:**
- <code>Z</code></code> represents the total cost of the diet.
- <code>x<sub>1</sub>, x<sub>2</sub>, x<sub>3</sub>, x<sub>4</sub>, x<sub>5</sub></code> are the decision variables, representing the number of servings of each food item.
- <code>c<sub>1</sub>, c<sub>2</sub>, c<sub>3</sub>, c<sub>4</sub>, c<sub>5</sub></code> are the costs per serving of each food item.

**Given these Constraints**
- <code>x<sub>1</sub>Na + x<sub>2</sub>Na + x<sub>3</sub>Na + x<sub>4</sub>Na + x<sub>5</sub>Na &le; 5000</code> ---Maximum allowed Sodium
- <code>x<sub>1</sub>Kc + x<sub>2</sub>Kc + x<sub>3</sub>Kc + x<sub>4</sub>Kc + x<sub>5</sub>Kc &ge; 2000</code> ---Minimum allowed Calories (kCal)
- <code>x<sub>1</sub>P  + x<sub>2</sub>P  + x<sub>3</sub>P +  x<sub>4</sub>P +  x<sub>5</sub>P  &ge; 50</code>   ------------Minimum allowed Protein
- <code>x<sub>1</sub>vD + x<sub>2</sub>vD + x<sub>3</sub>vD + x<sub>4</sub>vD + x<sub>5</sub>vD &ge; 20</code>   ----- Minimum allowed Vitamin D
- <code>x<sub>1</sub>Ca + x<sub>2</sub>Ca + x<sub>3</sub>Ca + x<sub>4</sub>Ca + x<sub>5</sub>Ca &ge; 1300</code> ---Minimum allowed Calcium
- <code>x<sub>1</sub>Fe + x<sub>2</sub>Fe + x<sub>3</sub>Fe + x<sub>4</sub>Fe + x<sub>5</sub>Fe &ge; 18</code>   ----- Minimum allowed Iron
- <code>x<sub>1</sub>K  + x<sub>2</sub>K  + x<sub>3</sub>K +  x<sub>4</sub>K +  x<sub>5</sub>K  &ge; 4700</code> --------- Minimum allowed Potassium 

The objective is to minimize <code>Z</code></code>, which is the sum of the costs of each food item multiplied by the number of servings of that food item. This represents the cost of the selected servings of food items in the diet.

The problem is finding the most cost-effective combination of food items that enable me to hit seven dietary requirements. 6 of those requirements are minimum thresholds with no upper bounds, meaning it's okay to exceed the values for Calories, Protein, Vitamin D, Calcium, Iron and Potassium. The only maximum threshold is for Sodium, which implies the opposite; having less Sodium than the threshold is okay, but I cannot have more. 
By looking at the combined value of each nutrient across all five food items, we can find the cheapest way to meet the required thresholds.

In [191]:
from pulp import LpProblem, LpMinimize, LpVariable, lpSum, LpStatus, value

# Create the LP problem
prob = LpProblem("Cost Minimization", LpMinimize)

#assign variables 
x1 = LpVariable("Eggs", 0, None)          # Single Serving Egg 
x2 = LpVariable("Toast", 0, None)         # Single Serving Bread 
x3 = LpVariable("Avocado", 0, None)       # Single Serving Avocado 
x4 = LpVariable("Curry", 0, None)         # Single Serving Curry 
x5 = LpVariable("Salmon Salad", 0, None)  # Single Serving Salmon Salad 

#Cost Function  
prob += (0.16)*x1 + (0.17)*x2 + (0.42)*x3 + (2.20)*x4 + (5.93)*x5        #Problem to minimize

#constraints 
prob += 70 * x1 + 170 * x2 +   3.5 * x3 + 480.3 * x4 + 1065 * x5 <= 5000 #sodium  
prob += 70 * x1 + 110 * x2 +  80.5 * x3 + 342.3 * x4 +  374 * x5 >= 2000 #kcal 
prob +=  6 * x1 +   5 * x2 +     1 * x3 +  16.6 * x4 + 17.5 * x5 >= 50   #protein
prob +=  1 * x1 +   0 * x2 +     0 * x3 +     0 * x4 + 6.2  * x5 >= 20   #Vitamin D
prob += 30 * x1 +  26 * x2 +     6 * x3 +  42.7 * x4 +  136 * x5 >= 1300 #Calcium 
prob +=0.9 * x1 +   1 * x2 +   0.3 * x3 +   1.3 * x4 + 1.88 * x5 >= 18   #Iron 
prob += 70 * x1 + 100 * x2 + 243.7 * x3 + 570.4 * x4 +  565 * x5 >= 4700 #Potassium

#Cannot eat negative amounts of any of the food items
prob += x1 >= 0
prob += x2 >= 0
prob += x3 >= 0
prob += x4 >= 0
prob += x5 >= 0

prob.solve();
# Print the status of the solution
print("Status:", LpStatus[prob.status]);

#calculate totals per nutrition metric 
daily_sodium = 70 * value(x1)  + 170 * value(x2)  +   3.5 * value(x3)  + 480.3 * value(x4)  + 1065 * value(x5)  #sodium  
daily_cKal = 70 * value(x1)  + 110 * value(x2)  +  80.5 * value(x3)  + 342.3 * value(x4)  +  374 * value(x5)   #kcal 
daily_protein =  6 * value(x1)  +   5 * value(x2)  +     1 * value(x3)  +  16.6 * value(x4)  + 17.5 * value(x5)    #protein
daily_vitD =  1 * value(x1)  +   0 * value(x2)  +     0 * value(x3)  +     0 * value(x4)  + 6.2  * value(x5)    #Vitamin D
daily_calcium = 30 * value(x1)  +  26 * value(x2)  +     6 * value(x3)  +  42.7 * value(x4)  +  136 * value(x5)  #Calcium 
daily_iron =0.9 * value(x1)  +   1 * value(x2)  +   0.3 * value(x3)  +   1.3 * value(x4)  + 1.88 * value(x5)     #Iron 
daily_potassium = 70 * value(x1)  + 100 * value(x2)  + 243.7 * value(x3)  + 570.4 * value(x4)  +  565 * value(x5)   #Potassium

# Print the optimal solution
print("Optimal Exchanges:");
for variable in prob.variables():
   # if variable.varValue != 0:
     print(variable.name, "=", variable.varValue)

# Calculate daily cost
daily_cost = value(x1) * (0.16) + value(x2) * (0.17) + value(x3) * (0.42) + value(x4) * (2.20)+ value(x5) * (5.93)
print("Meal Cost Limits:", round(value(daily_cost), 2))
#Print out full results 
print("Sodium:", round(value(daily_sodium), 2))
print("Calories:", round(value(daily_cKal), 2))
print("Protein:", round(value(daily_protein), 2))
print("Vitamin D:", round(value(daily_vitD), 2))
print("Calcium:", round(value(daily_calcium), 2))
print("Iron:", round(value(daily_iron), 2))
print("Potassium:", round(value(daily_potassium), 2))

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/maccassin/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/_m/0mhxm8wn1f1bv4tmfv5cwzjw0000gp/T/37f7d6162a83446787a4677f90bb9ae3-pulp.mps timeMode elapsed branch printingOptions all solution /var/folders/_m/0mhxm8wn1f1bv4tmfv5cwzjw0000gp/T/37f7d6162a83446787a4677f90bb9ae3-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 17 COLUMNS
At line 60 RHS
At line 73 BOUNDS
At line 74 ENDATA
Problem MODEL has 12 rows, 5 columns and 37 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 7 (-5) rows, 5 (0) columns and 32 (-5) elements
0  Obj 0 Primal inf 145.13596 (6)
3  Obj 9.103252
Optimal - objective value 9.103252
After Postsolve, objective 9.103252, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 9.103251999 - 3 iterations time 0.002, Presolve 0.00
Option for 

This first run results in a very unrealistic diet of more than 26 eggs and 18 slices of bread per day, 0 curry, 0 salmon salad, and 4.1 servings of avocado, which happens to work out to one entire Hass avocado. This makes sense as those three items have the least Sodium and are relatively the cheapest options. With Sodium being the only upper threshold, we are considering any high Sodium items like the Salad or Curry would be limited. That being said, I like putting salt on my eggs, so I should incorporate that additional Sodium into the amount associated with consuming one egg.

From this first scenario, the daily cost for me to hit these nutritional guidelines with my listed ingredients is $9.10 or about $273 a month, an amount less than my current monthly grocery budget. 

Furthermore, when looking at the daily total for each nutrient, we can see that we hit the Maximum for Sodium and the minimum for Calcium and Potassium. I also overshot the caloric target by 2213.38 kcal, which would lead to long-term weight gain if maintained long-term. If I were to alter my diet based on this information, I would look for foods higher in calcium and potassium rather than changing the amount of these existing items. 

In [192]:
from pulp import LpProblem, LpMinimize, LpVariable, lpSum, LpStatus, value

# Create the LP problem
prob = LpProblem("Cost Minimization", LpMinimize)

#assign variables 
x1 = LpVariable("Eggs", 0, None)          # Single Serving Egg 
x2 = LpVariable("Toast", 0, None)         # Single Serving Bread 
x3 = LpVariable("Avocado", 0, None)       # Single Serving Avocado 
x4 = LpVariable("Curry", 0, None)         # Single Serving Curry 
x5 = LpVariable("Salmon Salad", 0, None)  # Single Serving Salmon Salad 

#Cost Function 
prob += (0.16)*x1 + (0.17)*x2 + (0.42)*x3 + (2.20)*x4 + (5.93)*x5         #Problem to minimize

#constraints 
prob += 70 * x1 + 170 * x2 +   3.5 * x3 + 480.3 * x4 + 1065 * x5 <= 5000 #sodium  
prob += 70 * x1 + 110 * x2 +  80.5 * x3 + 342.3 * x4 +  374 * x5 >= 2000 #kcal 
prob +=  6 * x1 +   5 * x2 +     1 * x3 +  16.6 * x4 + 17.5 * x5 >= 50   #protein
prob +=  1 * x1 +   0 * x2 +     0 * x3 +     0 * x4 + 6.2  * x5 >= 20   #Vitamin D
prob += 30 * x1 +  26 * x2 +     6 * x3 +  42.7 * x4 +  136 * x5 >= 1300 #Calcium 
prob +=0.9 * x1 +   1 * x2 +   0.3 * x3 +   1.3 * x4 + 1.88 * x5 >= 18   #Iron 
prob += 70 * x1 + 100 * x2 + 243.7 * x3 + 570.4 * x4 +  565 * x5 >= 4700 #Potassium

#added constraints
prob +=  185 * x1 + 0 * x2 +    0 * x3 + 43.1 * x4 +    40 * x5 <= 300 # Cholesterol
prob +=    0 * x1 + 0 * x2 +    5 * x3 + 94.6 * x4 +    20 * x5 >= 90  # Vitamin C
prob +=  0.1 * x1 + 0 * x2 + 10.5 * x3 + 71.5 * x4 + 133.9 * x5 >= 120 # Vitamin K
prob +=  239 * x1 + 0 * x2 + 73.3 * x3 + 1375 * x4 + 289.1 * x5 >= 900 # Vitamin A
prob += 22.3 * x1 + 0 * x2 + 40.5 * x3 + 62.1 * x4 +  30.0 * x5 >= 400 # Folate


prob += x1 >= 0
prob += x2 >= 0
prob += x3 >= 0
prob += x4 >= 0
prob += x5 >= 0

# Solve the problem
prob.solve();
# Print the status of the solution
print("Status:", LpStatus[prob.status]);

# Print the optimal solution
print("Optimal Exchanges:");
for variable in prob.variables():
   # if variable.varValue != 0:
     print(variable.name, "=", variable.varValue)

#calculate totals per nutrition metric 
daily_sodium = 70 * value(x1)  + 170 * value(x2)  +   3.5 * value(x3)  + 480.3 * value(x4)  + 1065 * value(x5)  #sodium  
daily_cKal = 70 * value(x1)  + 110 * value(x2)  +  80.5 * value(x3)  + 342.3 * value(x4)  +  374 * value(x5)   #kcal 
daily_protein =  6 * value(x1)  +   5 * value(x2)  +     1 * value(x3)  +  16.6 * value(x4)  + 17.5 * value(x5)    #protein
daily_vitD =  1 * value(x1)  +   0 * value(x2)  +     0 * value(x3)  +     0 * value(x4)  + 6.2  * value(x5)    #Vitamin D
daily_calcium = 30 * value(x1)  +  26 * value(x2)  +     6 * value(x3)  +  42.7 * value(x4)  +  136 * value(x5)  #Calcium 
daily_iron =0.9 * value(x1)  +   1 * value(x2)  +   0.3 * value(x3)  +   1.3 * value(x4)  + 1.88 * value(x5)     #Iron 
daily_potassium = 70 * value(x1)  + 100 * value(x2)  + 243.7 * value(x3)  + 570.4 * value(x4)  +  565 * value(x5)   #Potassium
daily_cholesterol =  185 * value(x1)  + 0 * value(x2)  +    0 * value(x3)  + 43.1 * value(x4)  +    40 * value(x5)  # Cholesterol
daily_vitC =    0 * value(x1)  + 0 * value(x2)  +    5 * value(x3)  + 94.6 * value(x4)  +    20 * value(x5)   # Vitamin C
daily_vitK =  0.1 * value(x1)  + 0 * value(x2)  + 10.5 * value(x3)  + 71.5 * value(x4)  + 133.9 * value(x5)   # Vitamin K
daily_vitA =  239 * value(x1)  + 0 * value(x2)  + 73.3 * value(x3)  + 1375 * value(x4)  + 289.1 * value(x5)   # Vitamin A
daily_folate = 22.3 * value(x1)  + 0 * value(x2)  + 40.5 * value(x3)  + 62.1 * value(x4)  +  30.0 * value(x5)  # Folate

# Calculate daily cost
daily_cost = value(x1) * (0.16) + value(x2) * (0.17) + value(x3) * (0.42) + value(x4) * (2.20)+ value(x5) * (5.93)
#Print out full results 
print("Sodium:", round(value(daily_sodium), 2))
print("Calories:", round(value(daily_cKal), 2))
print("Protein:", round(value(daily_protein), 2))
print("Vitamin D:", round(value(daily_vitD), 2))
print("Calcium:", round(value(daily_calcium), 2))
print("Iron:", round(value(daily_iron), 2))
print("Potassium:", round(value(daily_potassium), 2))
print("Cholesterol:", round(value(daily_cholesterol), 2))
print("Vitamin C:", round(value(daily_vitC), 2))
print("Vitamin K:", round(value(daily_vitK), 2))
print("Vitamin A:", round(value(daily_vitA), 2))
print("Folate:", round(value(daily_folate), 2))

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/maccassin/.pyenv/versions/3.10.6/lib/python3.10/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/_m/0mhxm8wn1f1bv4tmfv5cwzjw0000gp/T/50d7c6c8cb824f64b888435a2abc476f-pulp.mps timeMode elapsed branch printingOptions all solution /var/folders/_m/0mhxm8wn1f1bv4tmfv5cwzjw0000gp/T/50d7c6c8cb824f64b888435a2abc476f-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 22 COLUMNS
At line 83 RHS
At line 101 BOUNDS
At line 102 ENDATA
Problem MODEL has 17 rows, 5 columns and 55 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 10 (-7) rows, 5 (0) columns and 41 (-14) elements
0  Obj 36.730727 Primal inf 20.4324 (2)
5  Obj 65.711283
Optimal - objective value 65.711283
After Postsolve, objective 65.711283, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 65.71128332 - 5 iterations time 0.002, Presolve 0.0

By adding constraints for Cholesterol, Vitamin C, Vitamin A, Vitamin K, and Folate, the most cost-effective diet has changed drastically in composition and price. 
The cost per day has gone from $9.10 to $65.71. That is a per-month change from $273.00 to $1971.30, which is well outside of my monthly grocery budget. 
For composition, I would now eat 109.7 servings of avocado (roughly 27.5 whole avocados), 0.95 of an egg, 3.07 servings of Salmon Salad, and 7.5 slices of toast. 

I would also consume over 10,000 kcal daily, which is unhealthy long-term. Based on the daily totals, I am still hitting the maximum for Sodium and now hitting the maximum for Cholesterol. I am still only hitting the minimum for Calcium, though now Potassium is above the minimum, and I am not hitting the minimum requirement for Vitamin D. If I were to work on changing my diet to hit these targets, I would look for foods higher in Calcium and Vitamin D. 

The addition of more constraints makes this problem more realistic. However, the lack of input options, i.e. foods I can eat, severely limits the practicality of implementing this problem in my day-to-day life. If constraints were shifted and focused on energy balance and macro-nutrients, this would be a handy tool for ensuring optimal training for endurance performance. Ensuring you are adequately fueled but maintaining a low weight is critical to peak performance. 